grakn-graql/src/main/java/ai/grakn/graql/internal/query/InsertQueryExecutor.java

```diff
@@ -67,285 +63,321 @@
      */
-    Answer insertAll() {
-        return insertAll(new QueryAnswer());
+    static Answer insertAll(Collection<VarPatternAdmin> patterns, GraknGraph graph) {
+        return create(patterns, graph).insertAll(new QueryAnswer());
     }
 
     /**
      * Insert all the Vars
      * @param results the result of a match query
      */
-    Answer insertAll(Answer results) {
-        concepts.clear();
-        concepts.putAll(results.map());
-        namedConcepts.clear();
-        namedConcepts.putAll(results.map());
-        vars.forEach(this::insertVar);
-        return new QueryAnswer(namedConcepts);
+    static Answer insertAll(Collection<VarPatternAdmin> patterns, GraknGraph graph, Answer results) {
+        return create(patterns, graph).insertAll(results);
     }
 
-    /**
-     * @param var the {@link VarPatternAdmin} to insert into the graph
-     */
-    private void insertVar(VarPatternAdmin var) {
-        Concept concept = getConcept(var);
-        var.getProperties().forEach(property -> ((VarPropertyInternal) property).insert(this, concept));
+    private static InsertQueryExecutor create(Collection<VarPatternAdmin> patterns, GraknGraph graph) {
+        ImmutableSet<VarAndProperty> properties =
+                patterns.stream().flatMap(VarAndProperty::fromPattern).collect(toImmutableSet());
+
+        /*
+            We build several many-to-many relations, indicated by a `Multimap<X, Y>`. These are used to represent
+            the dependencies between properties and variables.
+
+            `propDependencies.containsEntry(prop, var)` indicates that the property `prop` cannot be inserted until
+            the concept represented by the variable `var` is created.
+
+            For example, the property `$x isa $y` depends on the existence of the concept represented by `$y`.
+         */
+
+        Multimap<VarAndProperty, Var> propDependencies =
+                properties.stream().collect(flatteningToMultimap(identity(), VarAndProperty::requiredVars));
+
+        /*
+            `varDependencies.containsEntry(var, prop)` indicates that the concept represented by the variable `var`
+            cannot be created until the property `prop` is inserted.
+
+            For example, the concept represented by `$x` will not exist before the property `$x isa $y` is inserted.
+         */
+        Multimap<Var, VarAndProperty> varDependencies = HashMultimap.create();
+
+        Multimap<VarAndProperty, Var> producedVars =
+                properties.stream().collect(flatteningToMultimap(identity(), VarAndProperty::producedVars));
+
+        Multimaps.invertFrom(producedVars, varDependencies);
+
+        /*
+            Equivalent vars are variables that must represent the same concept as another var.
+
+                 $x label movie, sub entity;
+                 $y label movie;
+                 $z isa $y;
+
+            In this example, `$z isa $y` must not be inserted before `$y` is. However, `$y` does not have enough
```

Comment:
I am not sure I follow here. You say:

> `$z isa $y` must not be inserted before `$y`

Why is that? I don't see anything depending on `$z` if anything it looks like $y need to be inserted before $z and due to:

> $y label movie;

It looks like it can be.